### PR TITLE
fix springMVC4 noSuchMethod

### PR DESCRIPTION
**Pull request by SummerOfServenteen**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix

- Related issues
#2659
#3337 
___
### Bug fix
- Bug description.

- How to fix?
how about this way.I'm not sure whether I used the witness class correctly.I just test some version of springMVC ,and it works well.please check.

### CODE REVIEW

**Comment by SummerOfServenteen** [comment]
if  the version of springMVC is lower than  4.2.0.release ,the ControllerForLowVersionInstrumentation will be used ,ohterwise the ControllerInstrumentation will be used.

**Comment by wu-sheng** [comment]
You don't add it into config file, how it works?

File: apm-sniffer/apm-sdk-plugin/spring-plugins/mvc-annotation-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/mvc/v4/define/ControllerForLowVersionInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+package org.apache.skywalking.apm.plugin.spring.mvc.v4.define;
+
+public class ControllerForLowVersionInstrumentation extends AbstractControllerInstrumentation {
+    public static final String WITNESS_CLASSES_LOW_VERSION = "org.springframework.web.context.support.ServletContextPropertyPlaceholderConfigurer";
+
+    public static final String ENHANCE_ANNOTATION = "org.springframework.stereotype.Controller";
+
+    @Override
+    protected String[] witnessClasses() {
+        return new String[]{WITHNESS_CLASSES, "org.springframework.cache.interceptor.DefaultKeyGenerator", WITNESS_CLASSES_LOW_VERSION};
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/spring-plugins/mvc-annotation-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/mvc/v4/define/ControllerForLowVersionInstrumentation.java]: Why `org.springframework.cache.interceptor.DefaultKeyGenerator` exists without class variable? And have `WITNESS_CLASSES_LOW_VERSION` too. Which one is working?

File: apm-sniffer/apm-sdk-plugin/spring-plugins/mvc-annotation-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/mvc/v4/define/ControllerInstrumentation.java [hunk 2]
```diff
@@ -21,9 +21,17 @@
...
-    @Override protected String[] getEnhanceAnnotations() {
-        return new String[] {ENHANCE_ANNOTATION};
+    @Override
+    protected String[] getEnhanceAnnotations() {
+        return new String[]{ENHANCE_ANNOTATION};
+    }
+
+    @Override
+    protected String[] witnessClasses() {
+        return new String[]{WITHNESS_CLASSES, "org.springframework.cache.interceptor.DefaultKeyGenerator", WITNESS_CLASSES_HIGH_VERSION};
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/spring-plugins/mvc-annotation-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/mvc/v4/define/ControllerInstrumentation.java]: Same as above

File: apm-sniffer/apm-sdk-plugin/spring-plugins/mvc-annotation-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/mvc/v4/define/ControllerForLowVersionInstrumentation.java (hunk 1 above)
Comment (wu-sheng) [reply apm-sniffer/apm-sdk-plugin/spring-plugins/mvc-annotation-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/mvc/v4/define/ControllerForLowVersionInstrumentation.java]: `org.springframework.cache.interceptor.DefaultKeyGenerator` exist in low and high instrumentation definitions, what is the point?

Comment (dmsolr) [reply apm-sniffer/apm-sdk-plugin/spring-plugins/mvc-annotation-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/mvc/v4/define/ControllerForLowVersionInstrumentation.java]: `ControllerForLowVersionConstructorInterceptor ` dose not seem to be refered. We need to override `getConstructorsInterceptPoints()` here, right?

**Comment by wu-sheng** [comment]
Any update?

**Comment by SummerOfServenteen** [comment]
sorry,I was a little busy with my work.I'll check this problem carefully tomorrow.

**Comment by SummerOfServenteen** [comment]
Should i pr with a new brach?This pr is messy because of my rebase

**Comment by wu-sheng** [comment]
Do the thing you think better. You don't need our permission😄

**Pull request closed without merging**